src/main/java/ir/sahab/nexus/plugin/tag/internal/TagEntityAdapter.java

```diff
@@ -0,0 +1,174 @@
+package ir.sahab.nexus.plugin.tag.internal;
+
+import com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx;
+import com.orientechnologies.orient.core.db.record.OIdentifiable;
+import com.orientechnologies.orient.core.index.OIndex;
+import com.orientechnologies.orient.core.metadata.schema.OClass;
+import com.orientechnologies.orient.core.metadata.schema.OClass.INDEX_TYPE;
+import com.orientechnologies.orient.core.metadata.schema.OType;
+import com.orientechnologies.orient.core.record.impl.ODocument;
+import com.orientechnologies.orient.core.sql.query.OSQLSynchQuery;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Map.Entry;
+import java.util.stream.Collectors;
+
+import org.sonatype.nexus.orient.OClassNameBuilder;
+import org.sonatype.nexus.orient.OIndexNameBuilder;
+import org.sonatype.nexus.orient.entity.IterableEntityAdapter;
+
+import ir.sahab.nexus.plugin.tag.internal.dto.AssociatedComponent;
+
+/**
+ * Provides persisting/reading tags from database.
+ */
+public class TagEntityAdapter extends IterableEntityAdapter<TagEntity> {
+
+    private static final String TYPE_NAME = "tag";
+
+    private static final String DB_CLASS = new OClassNameBuilder().type(TYPE_NAME).build();
+
+    private static final String NAME_FIELD = "name";
+    private static final String ATTRIBUTES_FIELD = "attributes";
+    private static final String COMPONENTS_FIELD = "components";
+    private static final String FIRST_CREATED_FIELD = "firstCreated";
+    private static final String LAST_UPDATED_FIELD = "lastUpdated";
+
+    private static final String COMPONENT_REPOSITORY_FIELD = "repository";
+    private static final String COMPONENT_GROUP_FIELD = "group";
+    private static final String COMPONENT_NAME_FIELD = "name";
+    private static final String COMPONENT_VERSION_FIELD = "version";
+
+    private static final String NAME_INDEX = new OIndexNameBuilder().type(DB_CLASS).property("name").build();
+
+    public TagEntityAdapter() {
+        super(TYPE_NAME);
+    }
+
+    @Override
+    protected void defineType(OClass type) {
```

Comment:
I think the relation must be reversed. each component can have multiple tags.
or at least we should have index on both component name and tag name.
a user should be able to get all tags of a component.
Let's discuss about it